### ISSUE TITLE
Declarative webhook not being triggered

### ISSUE THREAD

**Issue by vincaslt** [Shopify/shopify-app-template-remix#818]
# Issue summary

Before opening this issue, I have:

- [ ] Upgraded to the latest version of the `@shopify` packages
  - Affected `@shopify/shopify-*` package and version:
  - Node version:
  - Operating system:
- [ ] Set `{ logger: { level: LogSeverity.Debug } }` in my configuration
- [ ] Found a reliable way to reproduce the problem that indicates it's a problem with the package
- [ ] Looked for similar issues in this repository
- [ ] Checked that this isn't an issue with a Shopify API
  - If it is, please create a post in the [Shopify community forums](https://community.shopify.com/c/partners-and-developers/ct-p/appdev) or report it to [Shopify Partner Support](https://help.shopify.com/en/support/partners/org-select)

I'm trying to register a new webhook for `orders/paid` topic, and I'm using this Remix template. I have this in my shopify.app.toml file:

```
[webhooks]
api_version = "2024-07"

  [[webhooks.subscriptions]]
  topics = [ "app/uninstalled", "orders/paid" ]
  uri = "/webhooks"
  compliance_topics = [ "customers/data_request", "customers/redact", "shop/redact" ]
  ```

I deployed this config `shopify app deploy`, I uninstalled/reinstalled the app. In development `npm run dev -- --reset` the `orders/paid` webhook is not firing at all, but uninstall does fire. I receive an empty list of registered webhooks when I check registered webhooks: https://<mystore>.[myshopify.com/admin/api/2024-07/webhooks.json](http://myshopify.com/admin/api/2024-07/webhooks.json)

The old way of adding webhooks to config when initializing shopify client still works fine and the webhooks.json contains my registered webhook:

```
webhooks: {
    [WebhookSubscriptionTopic.OrdersPaid]: {
      deliveryMethod: DeliveryMethod.Http,
      callbackUrl: "/webhooks",
    },
  },
  hooks: {
    afterAuth: async ({ session }) => {
      await shopify.registerWebhooks({ session });
    },
  },
```

Is there something that I do wrong?

## Expected behavior

The webhook fires.

## Actual behavior

Webhook is not fired.

## Debug logs

```
// Paste any relevant logs here
```

**Comment by FynnMenkIT** [comment]
Hey maybe its the way you register your webhook in combination with the problem I had here: https://github.com/Shopify/shopify-app-template-remix/issues/796#issuecomment-2252418725

you should reset your App, when registering a new webhook. For me only the npm run dev -- reset did not quite work so I deinstalled and used the command `shopify app dev --reset` to get the new webhooks being registered.

**Comment by vincaslt** [comment]
No, it doesn't work for me at all, only the uninstalled webhook is consistently being delivered. All others aren't (product update, order fulfilled), even by following the steps you mentioned exactly. My settings:

```
[webhooks]
api_version = "2024-07"

  [[webhooks.subscriptions]]
  topics = [
  "app/uninstalled",
  "discounts/create",
  "orders/fulfilled",
  "products/update"
]
  uri = "/webhooks"
  ```

I notice that when I deploy settings get pushed with hardcoded URLs (e.g. `application_url`) and when I then start the app with they change, I wonder if it has something to do with it. I'm assuming shopify doesn't know about the changed URLs, unless maybe  there's some handling for dev environment.

**Comment by vincaslt** [comment]
OK, so riding on the back of my observation, I got it to work like this:

1. `npm run dev -- --reset`
2. uninstall the app if installed before
3. install the app
4. `shopify app deploy` in a new console window
5. hooks are firing (until next start, then repeat the steps again).

Is there a better way that doesn't require juggling URLs?

**Comment by paulomarg** [comment]
I don't believe you should have to `deploy` on every restart if your topics haven't changed. I'll raise this with the team to see if there is anything we can do here.

**Comment by paulomarg** [comment]
UPDATE: for now, you'll need to `deploy` before `dev` when developing webhooks, so the URLs are updated. I've forwarded this feedback to the team, and we're looking for ways to improve this.

Since there isn't a way for us to change that behaviour from this repo, I'm going to close the issue, but the team is aware of it!

**Comment by grundmanise** [comment]
Hi @paulomarg, it’s been a while since this issue was created, but I just wanted to check if you know whether there were any fixes to the process or if we’re still expected to deploy before dev?
